sos/report/plugins/mellanox_firmware.py

```diff
@@ -0,0 +1,141 @@
+# Copyright (C) 2023 Nvidia Corporation, Alin Serdean <aserdean@nvidia.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+from sos.report.plugins import Plugin, IndependentPlugin
+import os
+import time
+
+
+class MellanoxFirmware(Plugin, IndependentPlugin):
+
+    short_desc = 'Nvidia(Mellanox) firmware tools output'
+
+    plugin_name = "mellanox_firmware"
+    profiles = ('hardware', 'system')
+    packages = ('mst', 'mstflint')
+
+    MLNX_STRING = "Mellanox Technologies"
+
+    def check_enabled(self):
+        """
+        Checks if this plugin should be executed at all.
+        We will only enable the plugin if there is a
+        Mellanox Technologies network adapter
+        """
+        lspci = self.exec_cmd("lspci -D -d 15b3::0200")
+        return lspci['status'] == 0 and self.MLNX_STRING in lspci['output']
+
+    def collect(self):
+        if not self.get_option('allow_system_changes'):
+            self._log_info("Skipping mst/mlx cable commands as system changes"
+                           "would be made. Use --allow-system-changes to"
+                           "enable this collection.")
+            return
+
+        """
+        Run only if mft package is installed.
+        flint is available from the mft package.
+        """
+        co = self.exec_cmd('flint --version')
+        if co['status'] != 0:
+            return
+
+        co = self.collect_cmd_output('mst start')
+        if co['status'] != 0:
+            return
+
+        self.collect_cmd_output('mst cable add')
+        self.collect_cmd_output("mst status -v", timeout=10)
+        self.collect_cmd_output("mlxcables", timeout=10)
+        co = os.listdir("/dev/mst")
+        mlxcables = []
+        for device in co:
+            if 'cable' in device:
+                mlxcables.append(device)
+        for mlxcable in mlxcables:
+            self.collect_cmd_output(f"mlxcables -d {mlxcable} --DDM",
+                                    timeout=10)
+            self.collect_cmd_output(f"mlxcables -d {mlxcable} --dump",
+                                    timeout=10)
+        self.collect_cmd_output("mst stop", changes=True)
+
+    def setup(self):
+        # Get all devices which have the vendor Mellanox Technologies
+        devices = []
+        device_list = self.collect_cmd_output('lspci -D -d 15b3::0200')
+        """
+        Will return a string of the following format:
+        0000:08:00.0 Ethernet controller: Mellanox Technologies MT2892
+        Family
+        """
+        if device_list['status'] != 0:
+            # bail out if there no Mellanox PCI devices
+            return
```

Comment:
This is required over the `check_enabled` method in cases when user manually enables the plugin, ACK.